__Анализ с помощью SQL__

Ваша первая задача как аналитика — проанализировать базу данных.
В ней — информация о книгах, издательствах, авторах, а также пользовательские обзоры книг. Эти данные помогут сформулировать ценностное предложение для нового продукта.

Описание данных
Таблица `books`
Содержит данные о книгах:
- `book_id` — идентификатор книги;
- `author_id` — идентификатор автора;
- `title` — название книги;
- `num_pages` — количество страниц;
- `publication_date` — дата публикации книги;
- `publisher_id` — идентификатор издателя.

Таблица `authors`
Содержит данные об авторах:

- `author_id` — идентификатор автора;
- `author` — имя автора.

Таблица `publishers`
Содержит данные об издательствах:

- `publisher_id` — идентификатор издательства;
- `publisher` — название издательства;

Таблица `ratings`
Содержит данные о пользовательских оценках книг:

- `rating_id` — идентификатор оценки;
- `book_id` — идентификатор книги;
- `username` — имя пользователя, оставившего оценку;
- `rating` — оценка книги.

Таблица `reviews`
Содержит данные о пользовательских обзорах:

- `review_id` — идентификатор обзора;
- `book_id` — идентификатор книги;
- `username` — имя автора обзора;
- `text` — текст обзора.

In [1]:
# импортируем библиотеки 
import pandas as pd
from sqlalchemy import create_engine


In [2]:

# устанавливаем параметры
db_config = {'user': 'praktikum_student', # имя пользователя
'pwd': 'Sdf4$2;d-d30pp', # пароль
'host': 'rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net',
'port': 6432, # порт подключения
'db': 'data-analyst-final-project-db'} # название базы данных
connection_string = 'postgresql://{}:{}@{}:{}/{}'.format(db_config['user'],
 db_config['pwd'],
 db_config['host'],
 db_config['port'],
 db_config['db'])
# сохраняем коннектор
engine = create_engine(connection_string, connect_args={'sslmode':'require'}) 

In [3]:
#занрузим таблицу с книгами
books = '''
            SELECT * 
            FROM books
        '''
books = pd.io.sql.read_sql(books, con = engine) 
books.head()

book_id  author_id                                              title  \
0        1        546                                       'Salem's Lot   
1        2        465                 1 000 Places to See Before You Die   
2        3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3        4         82  1491: New Revelations of the Americas Before C...   
4        5        125                                               1776   

   num_pages publication_date  publisher_id  
0        594       2005-11-01            93  
1        992       2003-05-22           336  
2        322       2010-12-21           135  
3        541       2006-10-10           309  
4        386       2006-07-04           268

In [4]:
books.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 6 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   book_id           1000 non-null   int64 
 1   author_id         1000 non-null   int64 
 2   title             1000 non-null   object
 3   num_pages         1000 non-null   int64 
 4   publication_date  1000 non-null   object
 5   publisher_id      1000 non-null   int64 
dtypes: int64(4), object(2)
memory usage: 47.0+ KB


У нас 1000 строк(книг).

In [5]:
#загрузим таблицу с авторами
authors = '''
            SELECT * 
            FROM authors
        '''
authors = pd.io.sql.read_sql(authors, con = engine) 
authors.head()

author_id                          author
0          1                      A.S. Byatt
1          2  Aesop/Laura Harris/Laura Gibbs
2          3                 Agatha Christie
3          4                   Alan Brennert
4          5        Alan Moore/David   Lloyd

In [6]:
authors.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 636 entries, 0 to 635
Data columns (total 2 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   author_id  636 non-null    int64 
 1   author     636 non-null    object
dtypes: int64(1), object(1)
memory usage: 10.1+ KB


In [8]:
#занрузим таблицу с изданиями
publishers = '''
            SELECT * 
            FROM publishers
        '''
publishers = pd.io.sql.read_sql(publishers, con = engine) 
publishers.head()

publisher_id                          publisher
0             1                                Ace
1             2                           Ace Book
2             3                          Ace Books
3             4                      Ace Hardcover
4             5  Addison Wesley Publishing Company

In [9]:
publishers.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 340 entries, 0 to 339
Data columns (total 2 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   publisher_id  340 non-null    int64 
 1   publisher     340 non-null    object
dtypes: int64(1), object(1)
memory usage: 5.4+ KB


In [10]:
#занрузим таблицу с пользовательскими обзорами.
ratings = '''
            SELECT * 
            FROM ratings
        '''
ratings = pd.io.sql.read_sql(ratings, con = engine) 
ratings.head()

rating_id  book_id       username  rating
0          1        1     ryanfranco       4
1          2        1  grantpatricia       2
2          3        1   brandtandrea       5
3          4        2       lorichen       3
4          5        2    mariokeller       2

In [11]:
ratings.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6456 entries, 0 to 6455
Data columns (total 4 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   rating_id  6456 non-null   int64 
 1   book_id    6456 non-null   int64 
 2   username   6456 non-null   object
 3   rating     6456 non-null   int64 
dtypes: int64(3), object(1)
memory usage: 201.9+ KB


In [12]:
reviews = '''
            SELECT * 
            FROM reviews
        '''
reviews = pd.io.sql.read_sql(reviews, con = engine)
reviews.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2793 entries, 0 to 2792
Data columns (total 4 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   review_id  2793 non-null   int64 
 1   book_id    2793 non-null   int64 
 2   username   2793 non-null   object
 3   text       2793 non-null   object
dtypes: int64(2), object(2)
memory usage: 87.4+ KB


Имеем: 
- 1000 книг
- 636 авторов
- 340 издательств
- 2793 обзора пользователей
- 6456 рейтинг пользователя

## Посчитаем, сколько книг вышло после 1 января 2000 года;

In [13]:
query_1 = '''
            SELECT COUNT(book_id) as count_of_books
            FROM books 
            WHERE publication_date > '01-01-2000'
        '''
pd.io.sql.read_sql(query_1, con = engine)

count_of_books
0             819

819 книг вышло после 1 января 2000 года.

## Для каждой книги посчитаем количество обзоров и среднюю оценку;

In [14]:
query_2 = '''
            SELECT b.title,
                   b.book_id AS id,  
                   AVG(rat.rating) AS avg_rating, 
                   COUNT(DISTINCT rev.text) AS count_reviews  
            FROM books AS b
            LEFT JOIN ratings AS rat ON rat.book_id = b.book_id
            LEFT JOIN reviews AS rev ON rev.book_id = b.book_id
            GROUP BY b.title, b.book_id
            ORDER BY count_reviews DESC;    
        '''
pd.io.sql.read_sql(query_2, con = engine)

title   id  avg_rating  \
0                              Twilight (Twilight  #1)  948    3.662500   
1                                  Water for Elephants  963    3.977273   
2                                     The Glass Castle  734    4.206897   
3    Harry Potter and the Prisoner of Azkaban (Harr...  302    4.414634   
4    The Curious Incident of the Dog in the Night-Time  695    4.081081   
..                                                 ...  ...         ...   
995    Anne Rice's The Vampire Lestat: A Graphic Novel   83    3.666667   
996                            The Natural Way to Draw  808    3.000000   
997   The Cat in the Hat and Other Dr. Seuss Favorites  672    5.000000   
998                          Essential Tales and Poems  221    4.000000   
999  Disney's Beauty and the Beast (A Little Golden...  191    4.000000   

     count_reviews  
0                7  
1                6  
2                6  
3                6  
4                6  
..             ...  
995              0  
996              0  
997              0  
998              0  
999              0  

[1000 rows x 4 columns]

Больше всего обзоров на `Twilight`, но оценка средняя не самая высокая - 3,66

## Определим издательство, которое выпустило наибольшее число книг толще 50 страниц — так вы исключите из анализа брошюры;

In [19]:
query_3 = '''
            SELECT p.publisher AS publiс,
                   COUNT(b.book_id) AS count_books
            FROM publishers AS p
            LEFT JOIN books AS b ON p.publisher_id = b.publisher_id
            WHERE b.num_pages > 50
            GROUP BY publiс
            ORDER BY count_books DESC
            LIMIT 5;
        '''
pd.io.sql.read_sql(query_3, con = engine)

publiс  count_books
0             Penguin Books           42
1                   Vintage           31
2  Grand Central Publishing           25
3          Penguin Classics           24
4                    Bantam           19

`Penguin Books`- выпустил больше всего книг более 50 страниц

## Определим автора с самой высокой средней оценкой книг — учитывая только книги с 50 и более оценками;

In [21]:

query_4 = '''
            SELECT sub.author AS author,
                   AVG(sub.avg_rating) 
            FROM (SELECT a.author,
                         b.book_id,
                         AVG(rat.rating) AS avg_rating
                  FROM authors AS a  
                  LEFT JOIN books AS b ON a.author_id = b.author_id
                  LEFT JOIN ratings AS rat ON b.book_id = rat.book_id
                  GROUP BY a.author, b.book_id
                  HAVING COUNT(rat.rating) >= 50) AS sub
            GROUP BY author
            ORDER BY AVG(sub.avg_rating) DESC
            LIMIT 5;
         '''   
pd.io.sql.read_sql(query_4, con = engine)

author       avg
0         J.K. Rowling/Mary GrandPré  4.283844
1  Markus Zusak/Cao Xuân Việt Khương  4.264151
2                     J.R.R. Tolkien  4.258446
3                  Louisa May Alcott  4.192308
4                       Rick Riordan  4.080645

У автора `J.K. Rowling/Mary GrandPré` самая высокая средняя оценка книги, учитывая только книги с 50 и более оценками

## Посчитаем среднее количество обзоров от пользователей, которые поставили больше 50 оценок.

In [22]:

query5 = '''
            SELECT ROUND(AVG(sub.count_reviews)) AS avg_count_reviews 
            FROM (SELECT COUNT(review_id) AS count_reviews 
                  FROM reviews
                  WHERE username IN (SELECT username
                                     FROM ratings
                                     GROUP BY username
                                     HAVING COUNT(rating) > 50)
                  GROUP BY username) AS sub
         '''   
pd.io.sql.read_sql(query5, con = engine)

avg_count_reviews
0               24.0

Среднее количество обзоров от пользователей, которые поставили больше 50 оценок - 24.